common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java

```diff
@@ -0,0 +1,138 @@
+import java.nio.file.Paths;
+import java.util.function.Consumer;
+
+/**
+ * Monitors for modifications to the specified file using a polling-based mechanism and performs the
+ * specified action (in the form of a callback) upon detection of a file modification.
+ *
+ * Unlike {@link FileModificationEventWatcher}, which doesn't work for monitoring modifications to a file which is
+ * a symbolic link to another file, an object of this class works for such files as well. However, since this object
+ * uses a polling-based mechanism - unlike the event-based mechanism used by {@link FileModificationEventWatcher}, lists
+ * files on each iteration and compares the output, this object is highly likely to be less efficient than an object of
+ * {@link FileModificationEventWatcher} - CPU cycles-wise.
+ *
+ */
+@Slf4j
+public class FileModificationPollingMonitor implements FileModificationMonitor {
+
+    private final static int DEFAULT_POLL_INTERVAL = 10 * 1000; // 10 seconds
+
+    /**
+     * The path of file to watch.
+     */
+    private final Path pathOfFileToWatch;
+
+    /**
+     * The action to perform when the a modification is detected for the specified file {@code pathOfFileToWatch}.
+     */
+    private final Consumer<File> callback;
+
+    private FileAlterationMonitor monitor;
+
+    private int pollingInterval;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param fileToWatch the file to watch
+     * @param callback    the callback to invoke when a modification to the {@code fileToWatch} is detected
+     *
+     * @throws NullPointerException if either {@code fileToWatch} or {@code callback} is null
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     * @throws NullPointerException if either {@code fileToWatch}  or {@code callback} is null
+     */
+    public FileModificationPollingMonitor(@NonNull String fileToWatch, @NonNull Consumer<File> callback)
+            throws FileNotFoundException {
+        this(fileToWatch, callback, DEFAULT_POLL_INTERVAL);
+    }
+
+    @VisibleForTesting
+    FileModificationPollingMonitor(@NonNull String fileToWatch, @NonNull Consumer<File> callback, int pollingInterval)
+            throws FileNotFoundException {
+        this.pathOfFileToWatch = Paths.get(fileToWatch);
+        this.validateInput(this.pathOfFileToWatch, callback);
+        this.callback = callback;
+        this.pollingInterval = pollingInterval;
+    }
+
+    @Override
+    public void startMonitoring() {
+        Path fileName = pathOfFileToWatch.getFileName();
+        log.debug("fileName obtained from pathOfFileToWatch is [{}]", fileName);
+        if (fileName == null) {
+            throw new IllegalStateException("fileName is 'null'");
+        }
+
+        Path dirPath = pathOfFileToWatch.getParent();
+        log.debug("dirPath is [{}]", dirPath);
+        if (dirPath == null) {
```

Comment:
How can this be?

Reply:
According to [Path api docs](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html), 
> Returns the parent path, or null if this path does not have a parent.